### PR TITLE
Feature/report builder

### PR DESCRIPTION
**Pull request by AdityaKhatri**
## Changes

* Create report builder

Depends on: https://github.com/the-deep/server/pull/1357

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] codegen errors
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [ ] translations

### CODE REVIEW

File: app/Base/configs/routes.tsx [hunk 1]
```diff
@@ -233,6 +233,76 @@ const entryEditRoute = wrap({
...
+        return (
+            project.allowedPermissions.includes('CREATE_ENTRY')
+            || project.allowedPermissions.includes('UPDATE_ENTRY')
+            || project.allowedPermissions.includes('DELETE_ENTRY')
+        );
+    },
+});
+
+const publicReportViewRoute = wrap({
+    path: '/reports/:reportSlug/',
```
Comment (tnagorra) [inline app/Base/configs/routes.tsx]: The reportEditRoute and publicReportViewRoute only differ on the type of the url param type. Let's have completely separate urls.

File: app/components/Portal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { createPortal } from 'react-dom';
+
+export interface Props {
+    children: React.ReactNode;
+    element?: HTMLDivElement;
+}
+
+function Portal(props: Props) {
+    const {
+        children,
```
Comment (tnagorra) [inline app/components/Portal/index.tsx]: Let's add a NOTE why we are not using Portal from deep-ui

File: app/views/PublicReportView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+    const { reportSlug } = useParams<{ reportSlug: string | undefined }>();
+
+    const variables = useMemo(() => (
+        reportSlug ? { slug: reportSlug } : undefined
+    ), [reportSlug]);
+
+    const {
+        data: completeData,
+    } = useQuery<PublicReportSnapshotQuery, PublicReportSnapshotQueryVariables>(
```
Comment (tnagorra) [inline app/views/PublicReportView/index.tsx]: Let's handle error and pending states.

File: app/views/PublicReportView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            return undefined;
+        }
+        return {
+            pathName: removeDomain(reportUrl),
+        };
+    }, [completeData]);
+
+    const {
+        data: yearlyDataNode,
+    } = useQuery<PublicReportDetailsQuery, PublicReportDetailsQueryVariables>(
```
Comment (tnagorra) [inline app/views/PublicReportView/index.tsx]: Let's handle error and pending states.

File: app/views/AnalysisDashboard/ReportsModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                    title
+                    id
+                    slug
+                    latestSnapshot {
+                        publishedBy {
+                            displayName
+                            id
+                        }
+                        publishedOn
+                    }
```
Comment (tnagorra) [inline app/views/AnalysisDashboard/ReportsModal/index.tsx]: Don't we have id on latestSnapshot?

File: app/views/PublicReportView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+                file {
+                    name
+                    url
+                }
+            }
+            reportDataFile {
+                name
+                url
+            }
+        }
```
Comment (tnagorra) [inline app/views/PublicReportView/index.tsx:48]: Do we not have id for file and reportDataFile?

File: app/components/GalleryFileUpload/index.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        $data: FileUploadInputType!,
+    ) {
+        fileUpload(data: $data) {
+            errors
+            ok
+            result {
+                file {
+                    name
+                    url
+                }
```
Comment (tnagorra) [inline app/components/GalleryFileUpload/index.tsx:24]: Don't we have id for file?

File: app/components/GalleryFileUpload/index.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            value={null}
+            onChange={handleFileInputChange}
+            status={undefined}
+            overrideStatus
+            maxFileSize={100}
+            accept={acceptFileType}
+            disabled={loading || disabled}
+        >
+            <IoCloudUpload />
+        </FileInput>
```
Comment (tnagorra) [inline app/components/GalleryFileUpload/index.tsx]: Let's add title

File: app/components/report/ReportBuilder/FileUpload/index.tsx [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                                }
+                            }
+                        }
+                    }
+                    report
+                    type
+                }
+            }
+        }
+    }
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/FileUpload/index.tsx:75]: Let's check if we have missed `id` in these query objects

File: app/components/report/ReportBuilder/MetadataEdit/index.tsx [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+    const handleHeadingContentStyleChange = useFormObject<
+        'headingContentStyle', HeadingContentStyleFormType
+    >('headingContentStyle', onConfigChange, {});
+
+    const configurationError = getErrorObject(error?.configuration);
+
+    return (
+        <div className={_cs(className, styles.metadataEdit)}>
+            <TextInput
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/MetadataEdit/index.tsx]: Don't we need a NonFieldError?

File: app/components/report/ReportBuilder/ReportContainer/BackgroundEdit/index.tsx [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    const error = getErrorObject(riskyError);
+    const onFieldChange = useFormObject<
+        NAME, BackgroundStyleFormType
+    >(name, onChange, {});
+
+    return (
+        <div className={styles.backgroundEdit}>
+            <Heading size="extraSmall">Background</Heading>
+            <div className={styles.inputs}>
+                <NumberInput
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/BackgroundEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/BorderEdit/index.tsx [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    const options = data?.borderVariants?.enumValues as EnumOptions<
+        AnalysisReportBorderStyleStyleEnum
+    >;
+
+    return (
+        <div className={styles.borderEdit}>
+            {loading && <PendingMessage />}
+            <Heading size="extraSmall">Border</Heading>
+            <div className={styles.inputs}>
+                <ColorInput
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/BorderEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/ContainerStylesEdit/index.tsx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        <ExpandableContainer
+            className={className}
+            heading="Container"
+            headingSize="small"
+            spacing="compact"
+            contentClassName={styles.expandedBody}
+            withoutBorder
+        >
+            {additionalStylingSettings}
+            <BorderEdit
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/ContainerStylesEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    contentData: ContentDataType[] | undefined;
+    configuration: AnalysisReportContainerContentConfigurationType | undefined;
+    generalConfiguration: ConfigType | undefined;
+    contentDataToFileMap: ContentDataFileMap | undefined;
+}
+
+function Content(props: Props) {
+    const {
+        contentType,
+        configuration: configurationFromProps = {},
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
        configuration: configurationFromProps,
```

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+
+    const configuration = removeNull(configurationFromProps);
+
+    if (contentType === 'HEADING') {
+        const {
+            heading: {
+                content,
+                variant,
+                style,
+            } = {},
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
            },
```

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+                    className={_cs(
+                        styles.heading,
+                        styles.headingFour,
+                    )}
+                    style={resolveTextStyle(
+                        style?.content,
+                        generalConfiguration?.headingContentStyle?.h4,
+                    )}
+                >
+                    {content ?? 'Title goes here'}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
                    {content || 'Title goes here'}
```
We might need to think about empty string

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+            </h1>
+        );
+    }
+
+    if (contentType === 'TEXT') {
+        const {
+            text: {
+                content,
+                style,
+            } = {},
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
            },
```

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+                </ReactMarkdown>
+            </div>
+        );
+    }
+
+    if (contentType === 'URL') {
+        const {
+            url: {
+                url,
+            } = {},
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
            },
```

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+        );
+    }
+
+    if (contentType === 'IMAGE') {
+        const {
+            image: {
+                caption,
+                altText,
+                style,
+            } = {},
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx]:
```suggestion
            },
```

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+                                style?.caption,
+                                generalConfiguration?.imageContentStyle?.caption,
+                            )}
+                        >
+                            {caption}
+                        </div>
+                    )}
+                </div>
+            );
+        }
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/Content/index.tsx:174]: Let's also have a fallback preview for image here

File: app/components/report/ReportBuilder/ReportContainer/HeadingEdit/index.tsx [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+                headingSize="small"
+                spacing="compact"
+                contentClassName={styles.expandedBody}
+                defaultVisibility
+                withoutBorder
+            >
+                <TextInput
+                    value={value?.content}
+                    name="content"
+                    label="Heading"
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/HeadingEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/ImageEdit/index.tsx [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+                headingSize="small"
+                spacing="compact"
+                contentClassName={styles.expandedBody}
+                defaultVisibility
+                withoutBorder
+            >
+                <FileUpload
+                    onSuccess={onFileUpload}
+                    acceptFileType="image/*"
+                    disabled={disabled}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/ImageEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/PaddingEdit/index.tsx [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+    const error = getErrorObject(riskyError);
+
+    return (
+        <div className={styles.paddingEdit}>
+            <Heading size="extraSmall">Padding</Heading>
+            <div className={styles.inputs}>
+                <NumberInput
+                    className={styles.input}
+                    name="top"
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/PaddingEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/TextEdit/index.tsx [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+        'style', TextContentStyleFormType
+    >('style', onFieldChange, {});
+
+    return (
+        <div className={_cs(className, styles.textEdit)}>
+            <ExpandableContainer
+                heading="General"
+                headingSize="small"
+                spacing="compact"
+                contentClassName={styles.expandedBody}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/TextEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/TextElementsStylesEdit/index.tsx [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+        <div className={_cs(className, styles.textElementsStylesEdit)}>
+            {loading && <PendingMessage />}
+            {label && (
+                <Heading size="extraSmall">{label}</Heading>
+            )}
+            <div className={styles.inputs}>
+                <NumberInput
+                    className={styles.input}
+                    label="Size"
+                    value={value?.size}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/TextElementsStylesEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/ReportContainer/UrlEdit/index.tsx [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+                spacing="compact"
+                contentClassName={styles.expandedBody}
+                withoutBorder
+            >
+                <TextInput
+                    value={value?.url}
+                    label="Url"
+                    name="url"
+                    onChange={onFieldChange}
+                    error={error?.url}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/UrlEdit/index.tsx]: Let's add NonFieldError

File: app/components/report/ReportBuilder/index.tsx [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+        disabled,
+        organizationOptions,
+        onOrganizationOptionsChange,
+        contentDataToFileMap,
+        setContentDataToFileMap,
+        onContentEditChange,
+        leftContentRef,
+    } = props;
+
+    const [containerToEdit, setContainerToEdit] = useState<string>();
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/index.tsx:86]:
```suggestion
    const [containerToEdit, setContainerToEdit] = useState<string | undefined>();
```

File: app/components/report/ReportBuilder/index.tsx [hunk 28]
```diff
@@ -0,0 +1,256 @@
...
+                    {!readOnly && (
+                        <QuickActionButton
+                            name={undefined}
+                            onClick={handleContentEdit}
+                            className={styles.editButton}
+                            disabled={disabled}
+                        >
+                            <IoPencil />
+                        </QuickActionButton>
+                    )}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/index.tsx]: Let's add title

File: app/components/report/ReportBuilder/index.tsx [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+                </div>
+                <NonFieldError error={getErrorObject(error)?.containers} />
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/index.tsx:217]: Let's add NonFieldError for the root

File: app/components/report/ReportBuilder/index.tsx [hunk 30]
```diff
@@ -0,0 +1,256 @@
...
+                    errored={false}
+                    filtered={false}
+                    pending={false}
+                />
+            </div>
+            {containerToEdit === 'metadata' && leftContentRef?.current && (
+                <Portal element={leftContentRef.current}>
+                    <Container
+                        className={styles.editContainer}
+                        heading="Metadata Edit"
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/index.tsx]:
```suggestion
                        heading="Edit Metadata"
```

File: app/components/report/ReportBuilder/styles.css [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+
+    .report {
+        display: flex;
+        flex-direction: column;
+        margin: 0 auto;
+        box-shadow: rgba(0, 0, 0, 0.16) 0px 1px 4px;
+        background-color: #fff;
+        padding: var(--dui-spacing-extra-large);
+        width: 210mm;
+        overflow-y: auto;
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/styles.css]: Are we hard-coding these values?

File: app/components/report/Toc/index.tsx [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+        const elementToScrollTo = document.getElementById(containerIdForLabel);
+
+        if (elementToScrollTo) {
+            elementToScrollTo.scrollIntoView({ behavior: 'smooth', block: 'start' });
+            hideNavbar();
+        }
+    }, [
+        label,
+        hideNavbar,
+    ]);
```
Comment (tnagorra) [inline app/components/report/Toc/index.tsx:57]: The containerIdForLabel might not always be unique.

File: app/components/report/Toc/index.tsx [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+                    <Button
+                        className={_cs(
+                            styles.menu,
+                        )}
+                        name="toggle"
+                        variant="transparent"
+                        onClick={toggleNavVisibility}
+                    >
+                        <IoMenuOutline className={styles.icon} />
+                    </Button>
```
Comment (tnagorra) [inline app/components/report/Toc/index.tsx]: Let's add title

File: app/components/report/schema.ts [hunk 34]
```diff
@@ -0,0 +1,344 @@
...
+        : (
+            T extends (infer Z)[]
+                ? DeepReplace<Z, A, B>[]
+                : (
+                    T extends object
+                        ? { [K in keyof T]: DeepReplace<T[K], A, B> }
+                        : T
+                )
+        )
+)
```
Comment (tnagorra) [inline app/components/report/schema.ts:46]: Let's add a comment why we are adding these types here.

File: app/components/report/utils.ts [hunk 35]
```diff
@@ -0,0 +1,129 @@
...
+        padding,
+        border,
+        background,
+    } = containerStyle ?? {};
+
+    const {
+        padding: generalPadding,
+        border: generalBorder,
+        background: generalBackground,
+    } = generalContainerStyle ?? {};
```
Comment (tnagorra) [inline app/components/report/utils.ts:72]:
```suggestion
    if (!containerStyle && !generalContainerStyle) {
        return {};
    }

    const {
        padding,
        border,
        background,
    } = containerStyle;

    const {
        padding: generalPadding,
        border: generalBorder,
        background: generalBackground,
    } = generalContainerStyle;
```

File: app/components/report/utils.ts [hunk 36]
```diff
@@ -0,0 +1,129 @@
...
+        size,
+    } = textStyle ?? {};
+
+    const {
+        align: generalAlign,
+        color: generalColor,
+        family: generalFamily,
+        weight: generalWeight,
+        size: generalSize,
+    } = generalTextStyle ?? {};
```
Comment (tnagorra) [inline app/components/report/utils.ts:112]:
```suggestion
    } = textStyle;

    const {
        align: generalAlign,
        color: generalColor,
        family: generalFamily,
        weight: generalWeight,
        size: generalSize,
    } = generalTextStyle;
```

File: app/gqlFragments.ts [hunk 37]
```diff
@@ -262,10 +262,20 @@ export const ORGANIZATION_FRAGMENT = gql`
...
+        }
         mergedAs {
             id
             title
             shortName
+            logo {
+                file {
+                    url
+                }
+            }
```
Comment (tnagorra) [inline app/gqlFragments.ts]: Let's check if we need to fetch `id`

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 38]
```diff
@@ -0,0 +1,573 @@
...
+
+import styles from './styles.css';
+
+const widthSelector = (item: { width: number }) => item.width;
+export function reorder<T>(
+    data: T[],
+    orderKey = 'order',
+) {
+    return data.map((v, i) => ({ ...v, [orderKey]: i + 1 }));
+}
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: We might need to type the return type strictly.

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 39]
```diff
@@ -0,0 +1,573 @@
...
+        contentDataToFileMap,
+        setContentDataToFileMap,
+        disabled,
+        generalConfiguration,
+        isBeingEdited,
+        onContentEditChange,
+        leftContentRef,
+    } = props;
+
+    const index = allItems?.findIndex((item) => item.clientId === containerKey);
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: Let's memoize this.

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 40]
```diff
@@ -0,0 +1,573 @@
...
+        rowItems.findIndex((item) => item.column === column)
+    ), [rowItems, column]);
+
+    const totalColSpan = useMemo(() => (
+        sum(rowItems.map((item) => item.width).filter(isDefined))
+    ), [rowItems]);
+
+    const widthOptions = useMemo(() => (
+        Array.from({ length: (12 - (totalColSpan - width)) - 2 }, (_, i) => ({ width: i + 3 }))
+    ), [totalColSpan, width]);
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: Let's add a NOTE here

File: app/components/report/ReportBuilder/ReportContainer/index.tsx (hunk 38 above)
Comment (tnagorra) [reply app/components/report/ReportBuilder/ReportContainer/index.tsx]: We can also optimize the re-order function by not editing the object if the new order is equal to old order

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 41]
```diff
@@ -0,0 +1,573 @@
...
+                width: 6,
+            };
+            const indexOfCurrentRowBefore = oldVal.findIndex((item) => item.row === row);
+            newItems.splice(indexOfCurrentRowBefore, 0, newItem);
+            return newItems.map((item) => ({
+                ...item,
+                row: ((item.row ?? 0) >= newItem.row && item.clientId !== newItem.clientId)
+                    ? ((item.row ?? 0) + 1)
+                    : item.row,
+            }));
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: Let's add comment here.

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 42]
```diff
@@ -0,0 +1,573 @@
...
+                width: 6,
+            };
+            const indexOfRow = oldVal.findIndex((item) => item.row === row);
+            newItems.splice(indexOfRow + 1, 0, newItem);
+            return newItems.map((item) => ({
+                ...item,
+                row: ((item.row ?? 0) >= newItem.row && item.clientId !== newItem.clientId)
+                    ? ((item.row ?? 0) + 1)
+                    : item.row,
+            }));
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: Let's add comment here.

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 43]
```diff
@@ -0,0 +1,573 @@
...
+
+    const isErrored = analyzeErrors(error);
+
+    const containerStyles = resolveContainerStyle(style, generalConfiguration?.containerStyle);
+
+    const heading = useMemo(() => {
+        if (contentType !== 'HEADING') {
+            return undefined;
+        }
+        return configuration?.heading?.content?.replace(' ', '-');
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx:363]: Why are replacing space with dash?

File: app/components/report/ReportBuilder/ReportContainer/index.tsx [hunk 44]
```diff
@@ -0,0 +1,573 @@
...
+                            additionalStylingSettings={(
+                                <NumberInput
+                                    name="height"
+                                    label="Height"
+                                    value={height}
+                                    onChange={onFieldChange}
+                                    disabled={disabled}
+                                />
+                            )}
+                        />
```
Comment (tnagorra) [inline app/components/report/ReportBuilder/ReportContainer/index.tsx]: Don't we need to pass error here?

File: app/views/ReportEdit/index.tsx [hunk 45]
```diff
@@ -0,0 +1,804 @@
...
+                    filters
+                    clientId
+                    statements {
+                        clientId
+                        informationGaps
+                        statement
+                        reportText
+                    }
+                }
+            }
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: Let's check if we need to fetch `id` here

File: app/views/ReportEdit/index.tsx [hunk 46]
```diff
@@ -0,0 +1,804 @@
...
+                            },
+                        },
+                    };
+
+                    return ([
+                        ...acc,
+                        header,
+                        ...(item?.statements ?? []).map((statement, statementIndex) => ({
+                            clientId: randomString(),
+                            row: acc.length + 1 + statementIndex + 1,
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx:433]: How does this work?

File: app/views/ReportEdit/index.tsx [hunk 47]
```diff
@@ -0,0 +1,804 @@
...
+                            contentType: 'TEXT' as const,
+                            contentConfiguration: {
+                                text: {
+                                    content: [
+                                        '#### Main Statement',
+                                        statement.statement,
+                                        '#### Information Gaps',
+                                        statement.informationGaps,
+                                        '#### My Analysis',
+                                        statement.reportText,
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: Don't we need to check if we have content for statement, informationGaps and reportText?

File: app/views/ReportEdit/index.tsx [hunk 48]
```diff
@@ -0,0 +1,804 @@
...
+                    column: 1,
+                    width: 12,
+                    contentType: 'HEADING' as const,
+                    contentConfiguration: {
+                        heading: {
+                            content: response.project.analysis.title,
+                            variant: 'H1',
+                        },
+                    },
+                }] as ReportContainerType[]);
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: We should not cast the type here.

File: app/views/ReportEdit/index.tsx [hunk 49]
```diff
@@ -0,0 +1,804 @@
...
+                        containers: newContainers?.map((item) => ({
+                            ...item,
+                            contentData: item.contentData?.map((contentDataItem) => ({
+                                ...contentDataItem,
+                                upload: contentDataItem.upload.id,
+                            })),
+                        })),
+                    },
+                );
+                const uploadItems = valueToSet.containers?.map((item) => item.contentData).flat();
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: We can also directly use flatMap

File: app/views/ReportEdit/index.tsx [hunk 50]
```diff
@@ -0,0 +1,804 @@
...
+
+    const pending = analysisDataLoading
+        || reportLoading
+        || createReportLoading
+        || updateReportLoading;
+
+    const [
+        contentEditPaneVisible,
+        setContentEditPaneVisibility,
+    ] = useState(false);
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: Let's move this to the top.

File: app/views/ReportEdit/index.tsx [hunk 51]
```diff
@@ -0,0 +1,804 @@
...
+
+    const selectedReport = data?.project?.analysisReport;
+    const publishConfirmMessage = (compareDate(
+        selectedReport?.modifiedAt,
+        selectedReport?.latestSnapshot?.publishedOn,
+    )) > 0 ? (
+            'Are you sure you want to published the report?'
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]:
```suggestion
            'Are you sure you want to publish the report?'
```

File: app/views/ReportEdit/index.tsx [hunk 52]
```diff
@@ -0,0 +1,804 @@
...
+        ) : (
+            `Looks like the latest version of report is already published.
+            Are you sure you want to published the report?`
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]:
```suggestion
            Are you sure you want to publish the report?`
```

File: app/views/ReportEdit/index.tsx [hunk 53]
```diff
@@ -0,0 +1,804 @@
...
+                            onClick={createSubmitHandler(
+                                validate,
+                                setError,
+                                handleSubmit,
+                            )}
+                            disabled={pristine}
+                            variant="primary"
+                        >
+                            Save
+                        </Button>
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: Let's also disable the button when we are doing any action.

File: app/views/ReportEdit/index.tsx [hunk 54]
```diff
@@ -0,0 +1,804 @@
...
+                            )}
+                        />
+                    )}
+                    <ReportBuilder
+                        className={styles.reportBuilder}
+                        reportId={reportId}
+                        value={value}
+                        error={error}
+                        setFieldValue={setFieldValue}
+                        disabled={false}
```
Comment (tnagorra) [inline app/views/ReportEdit/index.tsx]: We should also disable the report builder while we are waiting for the mutation

**Review by tnagorra — DISMISSED** [review]

File: app/views/PublicReportView/index.tsx (hunk 6 above)
Comment (AdityaKhatri) [reply app/views/PublicReportView/index.tsx:48]: no

File: app/components/GalleryFileUpload/index.tsx (hunk 7 above)
Comment (AdityaKhatri) [reply app/components/GalleryFileUpload/index.tsx:24]: no

File: app/components/report/ReportBuilder/FileUpload/index.tsx (hunk 9 above)
Comment (AdityaKhatri) [reply app/components/report/ReportBuilder/FileUpload/index.tsx:75]: its all good

File: app/components/report/ReportBuilder/index.tsx (hunk 27 above)
Comment (AdityaKhatri) [reply app/components/report/ReportBuilder/index.tsx:86]: auto inferred

File: app/components/report/ReportBuilder/styles.css (hunk 31 above)
Comment (AdityaKhatri) [reply app/components/report/ReportBuilder/styles.css]: somewhat yes

File: app/components/report/Toc/index.tsx (hunk 32 above)
Comment (AdityaKhatri) [reply app/components/report/Toc/index.tsx:57]: yes, and I think thats fine, we cannot really fix this without adding other indicators

File: app/components/report/utils.ts (hunk 35 above)
Comment (AdityaKhatri) [reply app/components/report/utils.ts:72]: Fixed in a weird way, type is not auto inferred.

File: app/views/ReportEdit/index.tsx (hunk 46 above)
Comment (AdityaKhatri) [reply app/views/ReportEdit/index.tsx:433]: Each pillar will add one header row and multiple statement rows. The loop of reduce will treat one pillar at a time.

File: app/views/ReportEdit/index.tsx (hunk 53 above)
Comment (AdityaKhatri) [reply app/views/ReportEdit/index.tsx]: The actions are saved immediately to form so shouldn't pose any issues.

File: app/components/report/ReportBuilder/ReportContainer/Content/index.tsx (hunk 20 above)
Comment (tnagorra) [reply app/components/report/ReportBuilder/ReportContainer/Content/index.tsx:174]: Is this added?

File: app/views/ReportEdit/index.tsx (hunk 53 above)
Comment (tnagorra) [reply app/views/ReportEdit/index.tsx]: Don't we still send a request to server?

File: app/components/report/Toc/index.tsx (hunk 32 above)
Comment (tnagorra) [reply app/components/report/Toc/index.tsx:57]: What do you mean by indicators?

File: app/components/report/utils.ts (hunk 35 above)
Comment (tnagorra) [reply app/components/report/utils.ts:72]: Why wouldn't it be inferred?

File: app/views/ReportEdit/index.tsx (hunk 46 above)
Comment (tnagorra) [reply app/views/ReportEdit/index.tsx:433]: How does this math work? Still not very clear. We need to add NOTE on how this works.

File: app/components/report/ReportBuilder/ReportContainer/styles.css [hunk 55]
```diff
@@ -0,0 +1,125 @@
...
+
+    /*
+    .mask {
+        display: none;
+        position: absolute;
+        top: 0;
+        left: 0;
+        opacity: 0.5;
+        background-color: var(--dui-color-background-information);
+        width: 100%;
```
Comment (barshathakuri) [inline app/components/report/ReportBuilder/ReportContainer/styles.css:75]: Is this a useful code?

File: app/translate/index.tsx [hunk 56]
```diff
@@ -1,25 +1,7 @@
 import React from 'react';
-import { isDefined, isNotDefined } from '@togglecorp/fujs';
+import { isDefined } from '@togglecorp/fujs';
 import devLang from './dev-lang.json';
```
Comment (barshathakuri) [inline app/translate/index.tsx]:
```suggestion
import { isDefined } from '@togglecorp/fujs';

import devLang from './dev-lang.json';
```

File: app/views/ProjectEdit/index.tsx [hunk 57]
```diff
@@ -103,7 +103,7 @@ function ProjectEdit() {
                                         className={styles.shareButton}
                                         name="copy"
                                         variant="secondary"
-                                        title="Share project to other users"
+                                        title="Get sharable link for other users"
```
Comment (barshathakuri) [inline app/views/ProjectEdit/index.tsx:106]:
```suggestion
                                        title="Get shareable link for other users"
```

File: app/views/PublicReportView/index.tsx [hunk 58]
```diff
@@ -0,0 +1,394 @@
...
+import {
+    _cs,
+    isDefined,
+    listToMap,
+    compareNumber,
+} from '@togglecorp/fujs';
+
+import {
+    useQuery,
+    gql,
```
Comment (barshathakuri) [inline app/views/PublicReportView/index.tsx:14]:
```suggestion
} from '@togglecorp/fujs';
import {
    useQuery,
    gql,
```

File: app/views/ReportEdit/index.tsx [hunk 59]
```diff
@@ -0,0 +1,810 @@
...
+} from '#base/utils/errorTransform';
+
+import ReportBuilder from '#components/report/ReportBuilder';
+import Toc from '#components/report/Toc';
+import schema, {
+    type PartialFormType,
+    type ReportContainerType,
+} from '#components/report/schema';
+import { ContentDataFileMap } from '#components/report/utils';
+import styles from './styles.css';
```
Comment (barshathakuri) [inline app/views/ReportEdit/index.tsx:81]:
```suggestion
import { ContentDataFileMap } from '#components/report/utils';

import styles from './styles.css';
```

**Review by barshathakuri — DISMISSED** [review]

**Review by puranban — DISMISSED** [review]

**Review by subinasr — APPROVED** [review]

**Pull request merged**